rerun_cpp/src/rerun/component_descriptor.hpp

```diff
@@ -0,0 +1,129 @@
+            descriptor.archetype_name = archetype_name_;
+            return descriptor;
+        }
+
+        /// Unconditionally sets `archetype_name` to the given one.
+        ComponentDescriptor with_archetype_name(const char* archetype_name_) const {
+            ComponentDescriptor descriptor = *this;
+            descriptor.archetype_name = archetype_name_;
+            return descriptor;
+        }
+
+        /// Unconditionally sets `archetype_field_name` to the given one.
+        ComponentDescriptor with_archetype_field_name(
+            std::optional<std::string_view> archetype_field_name_
+        ) const {
+            ComponentDescriptor descriptor = *this;
+            descriptor.archetype_field_name = archetype_field_name_;
+            return descriptor;
+        }
+
+        /// Unconditionally sets `archetype_field_name` to the given one.
+        ComponentDescriptor with_archetype_field_name(const char* archetype_field_name_) const {
+            ComponentDescriptor descriptor = *this;
+            descriptor.archetype_field_name = archetype_field_name_;
+            return descriptor;
+        }
+
+        /// Sets `archetype_name` to the given one iff it's not already set.
+        ComponentDescriptor or_with_archetype_name(std::optional<std::string_view> archetype_name_
+        ) const {
+            if (this->archetype_field_name.has_value()) {
+                return *this;
+            }
+            ComponentDescriptor descriptor = *this;
+            descriptor.archetype_name = archetype_name_;
+            return descriptor;
+        }
+
+        /// Sets `archetype_name` to the given one iff it's not already set.
+        ComponentDescriptor or_with_archetype_name(const char* archetype_name_) const {
+            if (this->archetype_field_name.has_value()) {
+                return *this;
+            }
+            ComponentDescriptor descriptor = *this;
+            descriptor.archetype_name = archetype_name_;
+            return descriptor;
+        }
+
+        /// Sets `archetype_field_name` to the given one iff it's not already set.
+        ComponentDescriptor or_with_archetype_field_name(
+            std::optional<std::string_view> archetype_field_name_
+        ) const {
+            if (this->archetype_field_name.has_value()) {
+                return *this;
+            }
+            ComponentDescriptor descriptor = *this;
+            descriptor.archetype_field_name = archetype_field_name_;
+            return descriptor;
+        }
+
+        /// Sets `archetype_field_name` to the given one iff it's not already set.
+        ComponentDescriptor or_with_archetype_field_name(const char* archetype_field_name_) const {
+            if (this->archetype_field_name.has_value()) {
+                return *this;
+            }
+            ComponentDescriptor descriptor = *this;
+            descriptor.archetype_field_name = archetype_field_name_;
+            return descriptor;
+        }
```

Comment:
Note: I tried to not repeat myself so much, but anytime I try to do something any fancier than that, either clang and/or gcc complains or things just segfault. Whatever.